### PR TITLE
Support save model checkpoint by step frequency

### PR DESCRIPTION
**Pull request by PatrickkZ**
## Description

Implement `after_train_iter`  method in ModelCheckpoint callback to support saving model checkpoint by step frequency. 
### 1. Why the change?
related issue: https://github.com/intel-analytics/BigDL/issues/7105#issuecomment-1419996858

### 2. User API changes

Add two parameters when creating the ModelCheckpoint callback.
- **by_epoch**: save chekpoint by epoch or by iteration, default is True
- **interval**: The saving period. If `by_epoch=True`, interval indicates epochs, otherwise it indicates iterations. Default is -1, which means "never".

To save model checkpoint by step frequency, we can create a ModelCheckpoint callback as below:
```python
# create estimator
estimator = ...

# create callbacks
callbacks = [
    ModelCheckpoint(filepath=ckpt_save_path, # filepath should contains 'epoch' or 'iter' like this 'my/path/test-{iter:02d}'
    save_weights_only=True,
    by_epoch=False,
    interval=5)
]

# use callbacks when training
estimator.fit(data, batch_size=batch_size, epochs=epochs,
    callbacks=callbacks,
    feature_cols=["feature"],
    label_cols=["label"])
```

### 3. Summary of the change 

Implement `after_train_iter`  method in ModelCheckpoint callback to support saving model checkpoint by step frequency. 

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

**Comment by jenniew** [comment]
> 

I think if by_epoch=True(default value), and interval=-1(default value), we may need to save checkpoint at each epoch, which is the common action just like pytorch lightning modelcheckpoint.

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -57,12 +64,25 @@ def after_train_epoch(self, runner):
         be called during TRAIN mode.
         :param epoch:  Integer, index of epoch.
         """
+        if not self.by_epoch:
+            return
```
Comment (jenniew) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py:68]: We need to check self.interval, and only checkpoint at specific epochs.

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -30,6 +30,8 @@ class ModelCheckpoint(Callback):
     def __init__(self,
                  filepath=None,
                  save_weights_only=False,
+                 by_epoch=True,
+                 interval=-1,
                  ):
```
Comment (jenniew) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py:35]: Please add type hint for the init function

**Review by leonardozcm — APPROVED** [review]
LGTM

**Review by jenniew — APPROVED** [review]

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -123,8 +150,8 @@ def _format_checkpoint_name(cls, dirname, filename, stats=None):
             for group in groups:
                 name = group[1:]
 
-                if "epoch" not in name:
-                    warnings.warn("We only support filepath with {epoch} for now.")
+                if "epoch" not in name and "iter" not in name:
+                    warnings.warn("We only support filepath with {epoch} or {iter} for now.")
```
[the remark this answers is not in the document]
Comment (PatrickkZ) [reply python/orca/src/bigdl/orca/learn/pytorch/callbacks/model_checkpoint.py:154]: the file path should have the format {epoch} or {iter}
```python
# example
filepath='my/path/sample-{epoch:02d}'
filepath='my/path/sample-{iter}'
```

**Pull request merged**